File: validpalindromeii.js
//Objective is the same as 'Valid Palindrome', except this time you are allowed
//to remove up to 1 character to make it a valid palindrome

let string = 'abca'


//O(n) solution that uses a 2-pointer approach to check if a string is a palindrome.
//If the two ends don't match, try to skip that character and see if it is a palindrome.

let left = 0
let right = string.length - 1

while (left < right) {
    //Ends don't match, so skip this one element and go left/right 
    if (string.charAt(left) !== string.charAt(right)) {
        return isPalindrome(string, left + 1, right) || isPalindrome(string, left, right - 1)
    }

    left++
    right--
}

return true 

function isPalindrome(string, left, right) {
    while (left < right) {
        if (string.charAt(left++) !== string.charAt(right--)) {
            return false
        }
    }

    return true
}

File: validpalindromeii.java
//Java Solution

class Solution {
    public boolean validPalindrome(String s) {
        int i = 0;
        int j = s.length() - 1;
        
        while (i < j) {
            if (s.charAt(i) != s.charAt(j)) {
                return isValid(i + 1, j, s) || isValid(i, j - 1, s);
            }
            
            i++;
            j--;
        }
        
        return true;
    }
    
    public boolean isValid(int i, int j, String s) {
        while (i < j) {
            if (s.charAt(i++) != s.charAt(j--)) {
                return false;
            }
        }
        
        return true;
    }
}